# Результат ревью:

Я отмечаю адекватный уровень твоей программистской подготовки. Ты используешь продвинутые концепции языка Python, библиотеки pandas и одной из основных библиотек машинного обучения sckit-learn. Это показывает твой уровень владения инструментами анализа данных и machine learning.

С комментированием никаких проблем нет. Твоих объяснений достаточно для понимания твоими коллегами хода мыслей. Ссылки из брифа проекта в нужных местах также помогают следовать за развитием событий и дают полную информацию касаемо того, что будет происходить в том или ином разделе проекта.

Мне было просто проверять твой проект, потому что он чётко структурирован, с использованием оглавления. 

Зелёным цветом я отметил и прокомментировал удачные и элегантные решения, на которые ты можешь опираться в будущих проектах.

Жёлтым цветом я выделил то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас.

Красных комментариев, которые бы не позволяли мне принять твой проект в его нынешнем состоянии - нет. 

ПРОЕКТ ПРИНЯТ!

Подводя итог: отличная работа, молодец! Желаю тебе успехов в учёбе, желаю, чтобы проекты сдавались или сразу, или после первой итерации! Удачи!

Если же после факта принятия проекта у тебя возникнут дополнительные вопросы – я никуда не пропал. Мой адрес эл/почты: anepomnjashiy@gmail.com, имя в Пачке: anepomnjashiy, имя в Телеграмме: https://t.me/aleks_nep

p.s. я тебе на твой адрес эл/почты сейчас вышлю некоторые материалы из своей электронной библиотеки - надеюсь они будут тебе полезными в учёбе.</div>

Приветствую Вадим! 

Меня зовут Александр Непомнящий и я буду проверять твой проект. Предлагаю общаться на «ты» :))) Но, если это не удобно - мы вполне можем перейти на вы".

Вкратце обо мне: я работаю в сфере информационных технологий, аналитика (анализ) данных - это и мой рабочий функционал и личная, интересная мне область для саморазвития. Используемый стек: python, pandas, графические библиотеки matplotlib/seaborn/plotly. Сфера применения: маркетинговая аналитика в медицинской клинике (Клиника «Кивач», www.kivach.ru, республика Карелия). Если вдруг есть вопрос – что делает аналитик в коммерческой медицине – то вот короткий ответ. Самый главный вопрос, на который мы хотим получить ответ (и который в дальнейшем приведёт к перераспределению рекламного бюджета, плюс напрямую повлияет на разработку и предложению рынку новых лечебных программ) - это КТО ОН, НАШ КЛИЕНТ? Фактически это задача (на нашем внутреннем языке формулируется как построение и анализ «ПОРТРЕТА КЛИЕНТА»). Визуализируем и пытаемся придать смысл половозрастным данным наших клиентов, кто из какого региона и в какие времена года к нам приезжает. Учитываем количество приездов. Учитываем на какую программу этот конкретный клиент приехал в первый раз и как в дальнейшем меняются его предпочтения. Пытаемся построить модель оттока и спрогнозировать, когда клиент уйдёт в отток. Смотрим цифры (суммы трат клиентов на отдельных группах медицинских программ). Анализируем медицинские услуги, которые входят в путёвку, а также что, в каких количествах и на какую сумму клиент покупает дополнительно. Очень вкратце как-то так.

Теперь к сути: основная цель моей проверки - пройти по шагам, по логике твоего решения и обсудить те моменты, которые можно изменить, улучшить, довести до идеала. Такой формат работы двух программистов (аналитиков) ты или уже встречал(а), или встретишь на своей новой работе. Кто-то такой процесс называет "код-ревью", ну а как по мне, так это обмен опытом двух коллег, когда каждый из нас становиться лучше в процессе дискуссии.

Свои мысли, советы, комментарии и вопросы я буду оставлять в текстовых ячейках (markdown) ПОД твоим кодом. 
Для твоего удобства все эти заметки будут "цветными", как на примерах ниже:

"Зелеными" комментариями я буду отмечать твой элегантный ("красивый") код и/или удачные решения, на которые ты можешь смело опираться в будущих проектах.
<div class="alert alert-success">
<b>Отлично! 👍</b>

Отличная находка, элегантный код ...</div>

"Жёлтые" комментарии - это мои размышления, предложения, что-то, что на следующей итерации можно сделать по-другому. Твоя реакция на "жёлтые" комментарий необходима и подразумевает или правки, или реакцию или аргументированное согласие/не согласие. Я искренне верю, что наш двусторонний диалог под "желтым" комментарием - это то, от чего твой проект только выиграет. <div class="alert alert-warning">
<b>Комментарий 👉</b>

Здесь я бы предложил использовать .... Мои аргументы в пользу этого ...</div>

"Красные" комментарий будут появляться, когда что-то пойдёт не так: например, код будет выдавать ошибку. Или, когда в проекте не будет ответов на основные вопросы, ради которых весь анализ и выполнялся. "Красные" комментарии необходимо будет исправить, прежде чем я смогу принять проект.
<div class="alert alert-danger">
<b>Необходимо исправить ❌</b>

В этой ячейке код работает не так как задумано ...</div>

Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Это поможет мне оперативно выполнить повторную проверку твоего проекта.

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом, лучше всего в новой ячейке, сразу под моим комментарием. Если для своих комментариев ты будешь использовать какой-то заметный цвет, то мне будет легче отследить твои изменения, например так:
<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Текст твоего комментария ...</font>
</div>

**Распространенные типичные недочёты в проектах студентов:**

<div class="alert alert-danger" ❌>
Неработоспособный код. Проверь проект на работоспособность перед отправкой на проверку. Это важный навык, который тебе пригодится как на собеседовании, при выполнении тестового задания, так и перед совещанием у руководства компании.</div>
<div class="alert alert-danger" ❌>
Проект не структурирован. В проекте должна быть структура, т.е. названия разделов и подразделов должны быть записаны в ячейках Markdown.</div>
<div class="alert alert-warning" 👉 >
Отсутствие описания, целей проекта, его задач, выводов. В описании должны быть озвучены поставленные задачи, содержание датасета, промежуточные и финальные выводы. Как правило, заказчик читает только финальный вывод, поэтому важно добавить информацию о проведенном исследовании, применить форматирование текста, в общем, сделать красиво, чтобы было понятно, что проведено настоящее исследование и выводу можно верить.</div>  
<div class="alert alert-warning" 👉 >
"Длинный код". Код и и комментарии, которые прячутся за правым концом ячейки нужно переносить на новую строку. Твои будущие коллеги не оценят необходимость постоянного скроллинга. Согласно PEP8: https://www.python.org/dev/peps/pep-0008/ длина строки не должна превышать 79 символов.</div>
<div class="alert alert-warning" 👉 >
Твои коллеги ценят и твое и своё время, поэтому "подсвечивай" свои комментарии в ячейках Markdown. Подробнее про синтаксис Markdown, с примерами, можно почитать здесь: https://paulradzkov.com/2014/markdown_cheatsheet/?
</div>

In [3]:
pip install fast_ml

     |████████████████████████████████| 42 kB 198 kB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [4]:
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from fast_ml.model_development import train_valid_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import f1_score
from sklearn.utils import shuffle
from sklearn.metrics import roc_auc_score

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Предобработка данных</a></span></li><li><span><a href="#Подготовка-выборок-для-моделей" data-toc-modified-id="Подготовка-выборок-для-моделей-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Подготовка выборок для моделей</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Отток клиентов

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

## Подготовка данных

### Предобработка данных

In [5]:
df = pd.read_csv('/datasets/Churn.csv')

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Отлично. Импортировали требуемые для анализа библиотеки и загрузили данные. Сейчас посмотрим на их состояние и если всё ОК, сможем приступать к построению моделей....
</div>

In [6]:
df.head()

RowNumber  CustomerId   Surname  CreditScore Geography  Gender  Age  \
0          1    15634602  Hargrave          619    France  Female   42   
1          2    15647311      Hill          608     Spain  Female   41   
2          3    15619304      Onio          502    France  Female   42   
3          4    15701354      Boni          699    France  Female   39   
4          5    15737888  Mitchell          850     Spain  Female   43   

   Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0     2.0       0.00              1          1               1   
1     1.0   83807.86              1          0               1   
2     8.0  159660.80              3          1               0   
3     1.0       0.00              2          0               0   
4     2.0  125510.82              1          1               1   

   EstimatedSalary  Exited  
0        101348.88       1  
1        112542.58       0  
2        113931.57       1  
3         93826.63       0  
4         79084.10       0

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Да, верно, info() показывает, что лишь в одном столбце данных нашего датафрейма есть пропуски. Ну что же, всего один столбец с пропусками лучше, чем целый пул ... задача чуть упрощается.
</div>

С **типами данных** все в порядке.  
Обнаружены **пустые значения** в столбце Tenure. Можно заполнить их медианным значением.

In [8]:
df.loc[df['Tenure'].isna(), 'Tenure'] = df['Tenure'].median()

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Отлично, молодец! Именно в этом проекте заполнение пропусков медианой - верное решение. Над другим способом, например заполнением чем-либо (очевидно по нескольким столбцам!?) пришлось бы серъёзно поразмыслить. В общем твой вариант обработки пропусков я считаю верным. 
    
Ну и как вариант, для расширения кругозора, какие варианты заполнения пропусков **я встречал в этом проекте у коллег** и почему они неверны:

1. Я не подтверждаю корректность заполнения пропусков в **Tenure - случайными** числами. Это как? Из каких соображений? Мы занимаемся наукой о данных (data science), а не случайными предсказаниями. 
  
2. я не подтверждаю корректность заполнения пропусков в **Tenure** фиксированной отрицательной "заглушкой". Во-первых, почему таким значением? Во-вторых, когда модель будет "проверять" новых, неизвестных ей заемщиков (чтобы предсказать "отток / не отток") - в этих новых строках будут отрицательные значения? Думаю нет, не будут, скорее там будут пропуски (судя по тем данным на которых мы сейчас будем модель строить).
  
3. я не подтверждаю корректность заполнения пропусков в **Tenure** фиксированной "заглушкой" (часто 0 выбирают). Я задаю вопрос: а почему скажем не 1 ... или 2 или 3? Ведь по смыслу данных, когда количество клиентов "Бета"-банка немного уменьшается с каждым последующим годом" (проверяется легко, запуском value_counts()) эти "заглушки" также походят по смыслу;

4. я не подтверждаю корректность заполнения пропусков в **Tenure** методами ffill / bfill. Смотри, подобный метод заполнения имеет право на существование, но мне кажется не в нашем случае. Скорее я использовал бы этот вариант для заполнения пропущенных значений в данных, когда значения (в данных) имеют временную или пространственную зависимость. Например, если данные представляют собой временной ряд (например ежедневный рост курса валют) или если данные связаны с географическими координатами. В таких случаях методы ffill и bfill могут быть более эффективными, чем заполнение пропущенных значений медианой или средним.
    
И ЕЩЁ СОВЕТ (применим, когда заполняем пропуски медианой, как в нашем случае): смотрим на статистики данных ДО заполнения и ПОСЛЕ заполнения, подтверждая таким образом, что наш метод заполнения не изменил распределение (статистики) в заполненном столбце.    
</div>

<div class="alert alert-warning">
<b>Комментарий 👉</b>

По результатам проверки примерно 2,5 тысяч проектов за крайние 12 месяцев у меня родилось небольшое эссе на тему "удаления vs. спасения" данных, как раз на основе тех ситуаций и вопросов, которые мне коллеги задавали, которые мы обсуждали, уточняли и исправляли ... 

Желание уменьшить количество пропусков в данных (если такое желание конечно возникает ... но в рабочих проектах, "в жизни" оно как правило возникает) - можно только приветствовать. Приветствовать по 3-м минимум причинам:

1. это тренировка
    
2. это попытка сохранить данные ("данных много не бывает")
    
3. самое главное на мой взгляд - развитие аналитического мышления, желания "проникать" в секреты, которые скрывают данные. Когда вначале рождается гипотеза, потом эта гипотеза проверяется на данных (может ли такое быть? подтверждается?) и в финале уже гипотеза облекается в код.

Однако заполнение пропусков в данных не такая простая задача, как кажется ... простого ответа на вопросы: заполнять? удалить? чем заполнять? подходящего под все сценарии, нет ни в одной умной книжке ...

Самый идеальный случай - когда у тебя под рукой есть эксперт в предметной области. Если такой есть - советуйся с ним. Как правило это человек из бизнеса, который знает суть бизнес-процесса и вот именно он даст тебе точный и аргументированный совет - что в данной, конкретной задаче делать с "битыми" полями в строках датафрейма. Ты скажешь ему какой процент пропусков в данных, а он тебе ответит что делать: можно/нельзя удалить и/или даст рекомендации по заполнению. Вообще, как правило такой человек "на любой работе" есть. Для примера, финансовый директор думаю всегда даст рекомендации по финансовым данным, прочие профильные спецы - по своей сфере компетенций. Наверное только в зарождающемся стартапе может не быть каких-то компетенций, но тут уже можно и через фриланс (и т.п.) попробовать найти спеца и просить совета.

Теперь, что делать, если такого человека нет. Я бы ответил так: **ПОПЫТАТЬСЯ** вникнуть в бизнес-проблему самому. Наверное это сложно, если предметная область ну совсем незнакома. Но в общем случае, если бизнес совсем незнаком, то заполнение пропусков в данных наверное, не получиться. В таком случае попробуй ориентироваться на такое правило: если записи с пропусками в ключевых для твоего анализа столбцах составляют 1-3-5-ну может до 10% строк, то попробуй эти записи из оригинального датафрейма вырезать (но сохрани бы в некий df, к которому ты при необходимости можешь всегда обратиться).

В жизни, МНЕ КАЖЕТСЯ, как правило по всему датафрейму заполнять медианой/средним - ну вообще не вариант. Почему? Я бы ответил так - потому, что реальные датафреймы имеют миллионы строк и горизонт в несколько (десятков) лет, и если в 90-2000-е года всё было как-то более -менее стабильно/прогнозируемо, то в нынешней реальности события раскручиваются гораздо быстрее и не прогнозируемо. Пример: ты аналитик в автобизнесе. У тебя есть данные с пропусками в цене для машин в 2005-2007 года. Можно заполнить их средними основываясь на данных 2020-2021?

Может быть и такой вариант, что без заполнения некоторых столбцов некоторые записи по сути **НЕ** имеют смысла, но если их (записи с пропусками удалить), то и данных-то не останется. Ну тогда я бы такой совет дал - всё же заполнять медианой (она к скачкам выбросов более устойчива), но не по всему глобальному датафрейму, а выбрать простые группы (например, группы по интервалу лет - чтобы сгладить турбулентность нынешних лет) или выбрать более сложные группы (например, группы по интервалу лет и одновременно марке автопроизводителя (это я про вымышленный пример с автобизнесом)).

Теперь предположим, что мы выбрали какую-то стратегию заполнения "пропусков", не суть какую (оставить NaN или заполнить "заглушкой"). И в том и в другом случае придётся помнить этот выбор. Ты спросишь почему? Отвечу - потому, что инструменты, применяемые тобой в Питон/Пандас в каких-то случаях могут спокойно пропускать твой выбор, а в каких-то случаях тебе придется этот твой выбор обрабатывать.

Пример 1: мы заполнили пропуски "заглушкой". В этом случае: 1. расчет корреляций значение "заглушку" будет "ловить" (надо отфильтровать!), 2. расчет гипотез также будет "сбивать" правильный вывод (надо отфильтровать!) 3. графики, которые ты будешь строить - будут "заглушку" отображать на осях (нужно фильтровать), 4. в группировках также по-видимому "заглушку" надо будет убирать.

Пример 2: мы оставили пропуски как NaN. В этом случае: 1. метод .corr NaN'ы пропускает (отфильтровать не надо), 2. в расчет гипотез всё равно пропуски отфильтровать 3. графики, которые ты будешь строить могут отображать, а могут не отображать, 4. в группировках NaN'ы по sum() видны не будут, а по count() будут.

Ну и напоследок совсем уже жёсткий пример: додумывать "от себя", бездумно проставлять в пропуски медианные/средние значения - это НЕ верно. Давай же опять "про жизнь". Я писал выше, что в медицине работаю, да? Предположим мне принесли данные из онкологического диспансера. Задача: построить модель, которая по некоторым маркерным анализам будет прогнозировать наличие онкологии. **По-простому, ответить, что стоит за строчкой в данных: жизнь или смерть.** Уууууух … аж мороз по коже продирает, правда? 

Итак, 1.000.000 записей и 70 столбцов. Данные за 25 лет. 70 столбцов - это поля, с результатами конкретных и очень специфичных анализов. В "начале" данных, 15-20 лет назад, есть пропуски (много), но и в принципе, в данных есть пропуски. Как думаешь - имею я моральное право перед людьми, которые фактически доверяют мне в виде данных свою жизнь огульно подойти к заполнению пропусков?
</div>

**Проверим на дубли**

In [9]:
df[df.duplicated()]

Empty DataFrame
Columns: [RowNumber, CustomerId, Surname, CreditScore, Geography, Gender, Age, Tenure, Balance, NumOfProducts, HasCrCard, IsActiveMember, EstimatedSalary, Exited]
Index: []

In [10]:
df[df.RowNumber.duplicated()]

Empty DataFrame
Columns: [RowNumber, CustomerId, Surname, CreditScore, Geography, Gender, Age, Tenure, Balance, NumOfProducts, HasCrCard, IsActiveMember, EstimatedSalary, Exited]
Index: []

In [11]:
df[df.CustomerId.duplicated()]

Empty DataFrame
Columns: [RowNumber, CustomerId, Surname, CreditScore, Geography, Gender, Age, Tenure, Balance, NumOfProducts, HasCrCard, IsActiveMember, EstimatedSalary, Exited]
Index: []

**Вывод:**
- Дублей не обнаружено
- типы столбцов корректные
- заполнены пропуски
- данные можно использовать

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Спасибо за точный и конкретный вывод по предварительному анализу загруженных данных. Зачастую коллеги ограничиваются исключительно кодом, без слов. Ввиду чего непонятно к каким выводам они пришли и что собираются делать дальше.

Ты важные моменты озвучиваешь в этом выводе, молодец. Я прокомментирую их по месту возникновения кода (или в финальном выводе по празделу подготовки (предобработки) данных.</div>

### Подготовка выборок для моделей

Такие данные, как фамилия клиента, номер строки, ID клиента, моделям ни о чем не скажут.

In [12]:
df.drop(columns=['RowNumber', 'CustomerId', 'Surname'], inplace=True)

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Здесь отлично. Обоснование почему эти столбцы в данных являются неинформативными, а соответственно и бесполезными для наших моделей машинного обучения ты верно приводишь. Всё так, эти столбцы можно "безболезненно" из нашего датафрейма удалить. Действительно, ведь **RowNumber** - дублирует индекс строки, начинает счет с единицы, **CustomerId** - уникальный идентификатор клиента, он не дает дополнительной полезной информации, **Surname** - фамилия - тоже идентификатор для определенного человека, но может повторятся, так как фамилии не уникальны, тоже бесполезен для исследования.
</div>

Категориальные текстовые данные кодируем методом OHE

In [13]:
df = pd.get_dummies(df, drop_first=True)

<div class="alert alert-warning">
<b>Комментарий 👉</b>

Сейчас в теории тренажера вам предлагается использовать get_dummies(), однако это приведёт к сложностям в будущих спринтах. Поэтому здесь я бы рекомендовал тебе использовать класс OneHotEncoder из sklearn (https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html).
        
Get_dummies подходит больше для анализа данных, а для машинного обучения более предпочтителен OneHotEncoding, т.к. он позволяет избежать ряд ошибок при обучении моделей, в том числе может работать с неизвестными ранее уровнями категорий, которых не было изначально (например, если появится еще одна страна, скажем Italy). Get_dummies с неизвестными категориями "из коробки" не работает, но есть лайфхак описанный здесь: https://pythonsimplified.com/difference-between-onehotencoder-and-get_dummies/ (*Though get_dummies can’t handle unknown categories natively, you could get around this inconsistency by applying the below technique. You will have to save the columns of the train set and load it during prediction on test set. Then you need to apply reindex and after filling the missing values will get you the same features as the train set*)
        
Пандасовский Get_dummies не имеет методов fit и transform, поэтому кодируем категории на полном датафрейме (ДО разбиения на выборки ... так мы кстати обходим и проблему "неизвестных/новых категорий"), затем "режем" данные на выборки.

НАСТОЯТЕЛЬНАЯ РЕКОМЕНДАЦИЯ в будущих проектах: используем OneHotEncoder из sklearn!
</div>

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Молодец, что сейчас ты не забываешь про dummy-ловушку.
</div>

In [14]:
# Масштабируем переменные

In [15]:
# колонки для масштабирования
# numerics = ['int64', 'float64', 'uint8']
# numeric_cols = df.select_dtypes(include=numerics).columns.tolist()
# numeric_cols.remove('Exited')
# numeric_cols

In [16]:
# # объявляем scaler
# scaler = StandardScaler()
# # обучаем
# df[numeric_cols] = scaler.fit_transform(df[numeric_cols])

<div class="alert alert-danger">
<b>Необходимо исправить ❌</b>

Здесь у тебя ошибка. Масштабирование на какой выборке должно "обучаться" а затем для каких выборок применяться?
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Перенес скалирование чуть ниже, после разделения на выборки. Неверный код выше закомментировал.</font>
</div>

<div class="alert alert-success">
<b>КОММЕНТАРИЙ V2</b>

Да, теперь верно!
</div>

Разделяем наши данные на 3 выборки

In [17]:
features_train, target_train, features_valid, target_valid, features_test, target_test = train_valid_test_split(
    df, target = 'Exited', train_size=0.7, valid_size=0.15, test_size=0.15, random_state=999)

print('train:', features_train.shape)
print('valid:', features_valid.shape)
print('test:', features_test.shape)

train: (7000, 11)
valid: (1500, 11)
test: (1500, 11)


In [18]:
# колонки для масштабирования
numerics = ['int64', 'float64', 'uint8']
numeric_cols = df.select_dtypes(include=numerics).columns.tolist()
numeric_cols.remove('Exited')
numeric_cols

['CreditScore',
 'Age',
 'Tenure',
 'Balance',
 'NumOfProducts',
 'HasCrCard',
 'IsActiveMember',
 'EstimatedSalary',
 'Geography_Germany',
 'Geography_Spain',
 'Gender_Male']

In [19]:
# # объявляем scaler
scaler = StandardScaler()
# обучаем
scaler.fit(features_train[numeric_cols])
# Масштабируем
features_train[numeric_cols] = scaler.transform(features_train[numeric_cols])
features_valid[numeric_cols] = scaler.transform(features_valid[numeric_cols])
features_test[numeric_cols] = scaler.transform(features_test[numeric_cols])

<div class="alert alert-success">
<b>КОММЕНТАРИЙ V2</b>



Последовательность команд (это **псевдокод**): 
    
    1. scaler.fit(X_train) 
    2. X_train_st = scaler.transform(X_train) 
    
можно было записать одной:
    
    X_train_st = scaler.fit_transform(X_train)
</div>

In [32]:
features_train.head()

CreditScore       Age    Tenure   Balance  NumOfProducts  HasCrCard  \
2572     0.620491  1.853015  1.452226  0.227444       0.801800   0.648202   
4540    -0.676503 -0.655623 -1.441170  1.168705      -0.911448   0.648202   
8384     1.077033 -0.462651 -0.717821 -1.240591       0.801800   0.648202   
3149    -0.085074 -0.462651  1.090552  0.648247      -0.911448   0.648202   
5479    -0.852895 -0.366165 -0.717821  0.640494      -0.911448  -1.542728   

      IsActiveMember  EstimatedSalary  Geography_Germany  Geography_Spain  \
2572        0.969882        -1.285115          -0.578230        -0.570532   
4540       -1.031053        -1.196694           1.729416        -0.570532   
8384        0.969882         1.021341          -0.578230        -0.570532   
3149       -1.031053        -0.610775          -0.578230         1.752751   
5479       -1.031053        -0.102248          -0.578230         1.752751   

      Gender_Male  
2572     0.918720  
4540     0.918720  
8384     0.918720  
3149    -1.088471  
5479     0.918720

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

1. Здесь хорошо, происходит фиксация random_state. Воспроизводимость результатов разбиения выборки на обучающую (тренировочную) / тестовую / валидационную мы обеспечили, значит при всех последующих запусках нашего кода подвыборки будут идентичными.
    
2. Доли размеров train/test/valid 70:15:15 - хорошо.
</div>

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Молодец, что после разбиения данных на выборки смотришь на их размеры и размерности. Метод shape для этого - идеальный помощник. "Цифры" по выборкам показывают верно ли мы произвели "разделение" данных.

p.s. так (**псевдокод**): X_train.shape[0] – «покажет» количество строк в тренировочной выборке, а X_train.shape[1] - количество столбцов в ней же. Ну а X_train.shape – выведет размерность train'а в виде кортежа с 2мя значениями (первое число – количество строк, второе – столбцов).</div>

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Данные загружены, отлично.

Хорошо, что применяешь метод .head()/.sample() таким образом в память не будет загружен весь датасет, объем которого может быть огромным.

Методы .info, .describe, .value_counts(), .unique() как правило часто встречаются и обязательны к применению в задачах по исследованию данных, ведь с их помощью можно получить первые инсайты. 

Итак, мы провели предварительный анализ данных, проверили данные на наличие пропусков и затем обработали их (аргументированно заполнили / удалили / оставили как есть). Теперь можно приступать собственно к анализу данных и ответам на вопросы проекта.</div>

<div class="alert alert-warning">
<b>Комментарий 👉</b>

Несколько моментов "на будущее" (сейчас забегаю чуть вперед, но будущем это будет важно) которые я хотел озвучить перед переходом к собственно исследованию:
    
- СДЕЛАНО_МОЛОДЕЦ имеет смысл критично посмотреть на данные с точки зрения наличия пропусков в признаках (вопрос: все ли ml-модели "любят" пропуски?) и обработать или удалить их 
    
- имеет смысл явно обозначить какую задачу мы будем решать: с учителем / без учителя, классификации / регрессии ...
    
- в моделях бинарной классификации имеет смысл смотреть на баланс классов. Два взгляда (с разных сторон) на эту проблему тут: https://alexanderdyakonov.wordpress.com/2021/05/27/imbalance/ и https://habr.com/ru/post/349078/
    
- имеет смысл смотреть на корреляцию признаков. При высокой их степени коррелированности возникает проблема  мультиколлиниарность признаков. Проблема "мультиколлинеарности" это и про переобучение модели и про её интерпретируемость и про сложность модели (содержание в модели избыточного количества переменных). В чем негативный эффект сложной модели? Вот в чем: усложняется интерпретация параметров, оценки параметров ненадежны – получаются большие стандартные ошибки, которые меняются с изменением объема наблюдений, что делает модель регрессии непригодной для прогнозирования.
    
А теперь давай про интерпретируемость. Согласись - это важно, чтобы результат, который выдает модель можно было бы понять и объяснить с точки зрения предметной области. В случае включения в модель коллинеарных признаков может случиться так, что коэффициенты (да, проблема присуща моделям линейным ... регрессия среди них!) могут менять знаки и попытка объяснения предметной области с помощью этих коэффициентов будет противоречить здравому смыслу.
    
Вот классная статья. Простым языком, на примере показана проблема. А также, что важно - разобраны варианты как её (проблему) победить. Статья: https://habr.com/ru/company/akbarsdigital/blog/592493/ 
</div>

## Исследование задачи

Посмотрим на наличие дисбаланса классов

In [22]:
df.Exited.value_counts()

0    7963
1    2037
Name: Exited, dtype: int64

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Отлично, молодец что обращаешь внимание на баланс классов в данных. Если в value_counts добавить normalize=True, то увидишь %, а не количество наблюдений по каждому классу.
</div>

**Вывод**: Дисбаланс определенно присутствует

Попробуем обучить модели с без учета дисбаланса классов и посчитать метрику F1

In [23]:
# DecisionTreeClassifier
best_f1 = 0
best_depth = 0
for depth in range(1, 16):
    model = DecisionTreeClassifier(random_state=999, max_depth=depth)
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > best_f1:
        best_f1 = f1
        best_depth = depth

print('Best depth:', best_depth, 'F1:', best_f1)

Best depth: 5 F1: 0.5962264150943397


<div class="alert alert-warning">
<b>Комментарий 👉</b>

Рекомендую перебрать в цикле 2-3 гиперпараметра. Одного, как для "дерева", так и "леса" мало - точность наверняка будет низкой. В итоге получишь сразу два плюса: потренируешся работать с несколькими гиперпараметрами одновременно и получишь более высокую точность моделей!
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>Понял, спасибо. Просто т.к. другие параметры в курсе мы вообще не затрагивали, не стал их пока использовать. Но в будущем попробую.</font>
</div>

<div class="alert alert-success">
<b>КОММЕНТАРИЙ V2</b>

+
</div>

In [24]:
# RandomForestClassifier
# сначала подбираем лучшую глубину
best_f1 = 0
best_depth = 0
for depth in range(1, 16):
    model = RandomForestClassifier(random_state=999, max_depth=depth)
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > best_f1:
        best_f1 = f1
        best_depth = depth

# по лучшей глубине подберем лучшее число деревьев
best_estims = 100 #by default
for estims in range(10, 161, 10):
    model = RandomForestClassifier(random_state=999, max_depth=best_depth, n_estimators=estims)
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > best_f1:
        best_f1 = f1
        best_estims = estims
        
print('Best depth:', best_depth, 'Best estims', best_estims, 'F1:', best_f1)

Best depth: 14 Best estims 100 F1: 0.6003976143141153


<div class="alert alert-warning">
<b>Комментарий 👉</b>

А почему не в двойном цикле?
</div>

<div style="background: #B0E0E6; padding: 5px; border: 1px solid SteelBlue; border-radius: 5px;">
    <font color='4682B4'><u><b>КОММЕНТАРИЙ СТУДЕНТА</b></u></font>
    <br />
    <font color='4682B4'>В теоретичской части эксперементировал с двойным циклом. Перебор был долгим, от 5 минут. Выяснил, что если сначала найти max_depth, то по нему уже можно искать n_estimators (в двойном цикле max_depth бы не изменился). По крайней мере, так было на тех данных. Судя по твоему комментарию, догадываюсь, что так не всегда :)</font>
</div>

<div class="alert alert-success">
<b>КОММЕНТАРИЙ V2</b>

Да нет, результат будет скорее всего тем же, просто визуально код будет состоять из 1 ячейки, а не из 2х. Симпатичнее будет выглядеть. 
</div>

In [25]:
# LogisticRegression
# подберем лучшее число итераций обучения
best_iters = 0 
best_f1 = 0
for iters in range(1, 20, 1):
    model = LogisticRegression(max_iter=iters, random_state=999, solver='lbfgs')
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > best_f1:
        best_f1 = f1
        best_iters = iters
        
print('Best iters', best_iters, 'F1:', best_f1)

Best iters 3 F1: 0.33185840707964603


**Вывод:**  
Лучшая модель без учета дибаланса классов - RandomForestClassifier с лучшим найденным `F1 = 0.6`

<div class="alert alert-warning">
<b>Комментарий 👉</b>

Вадим, я не буду комментировать далее рутинные (и практически одинаковые, за исключением меняющихся названий моделей, действия). Если только ошибку увижу - тогда напишу. А так, смотрю за ходом исследования + возможно где-то свои мысли/советы/рекомендации укажу.    
</div>

## Борьба с дисбалансом

Попробуем использовать гиперпараметр `class_weight='balanced'`

<div class="alert alert-warning">
<b>Комментарий 👉</b>

Отметь такой момент (сейчас этого недостатка в твоей реализации нет ... я просто рассуждаю): совмещать up(down)sampling и class_weight нельзя.

Взвешивание работает таким образом, что даёт дополнительный вес значимости малопредставленным объектам, жертвуя точностью предсказывания многопредставленного класса в пользу точности предсказывания малопредставленного. Увеличение (уменьшение) выборки делает по сути то же самое (дублирование выборки дает больший вес дублируемому классу). Таким образом, использование сразу двух методов балансирование привело бы новому дисбалансу, поэтому использовать такие методы нужно взаимоисключающе.
</div>

In [26]:
# DecisionTreeClassifier
best_f1 = 0
best_depth = 0
for depth in range(1, 16):
    model = DecisionTreeClassifier(random_state=999, max_depth=depth, class_weight='balanced')
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > best_f1:
        best_f1 = f1
        best_depth = depth

print('Best depth:', best_depth, 'F1:', best_f1)

Best depth: 7 F1: 0.5794621026894865


Для решающего дерева результат почему-то стал хуже, чем был

In [27]:
# RandomForestClassifier
# сначала подбираем лучшую глубину
best_f1 = 0
best_depth = 0
for depth in range(1, 16):
    model = RandomForestClassifier(random_state=999, max_depth=depth, class_weight='balanced')
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > best_f1:
        best_f1 = f1
        best_depth = depth

# по лучшей глубине подберем лучшее число деревьев
best_estims = 100 #by default
for estims in range(10, 161, 10):
    model = RandomForestClassifier(random_state=999, max_depth=best_depth, n_estimators=estims, class_weight='balanced')
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > best_f1:
        best_f1 = f1
        best_estims = estims
        
print('Best depth:', best_depth, 'Best estims', best_estims, 'F1:', best_f1)

Best depth: 9 Best estims 120 F1: 0.6505295007564296


RandomForestClassifier: F1 составил 0.65, что уже лучше против 0.6 для модели без учета несбалансированности классов

In [22]:
# LogisticRegression
# подберем лучшее число итераций обучения
best_iters = 0 
best_f1 = 0
for iters in range(1, 20, 1):
    model = LogisticRegression(max_iter=iters, random_state=999, solver='lbfgs', class_weight='balanced')
    model.fit(features_train, target_train)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > best_f1:
        best_f1 = f1
        best_iters = iters
        
print('Best iters', best_iters, 'F1:', best_f1)

Best iters 4 F1: 0.5147392290249434


LogisticRegression: результат стал значительно лучше.

Теперь сравним применение гиперпараметра `class_weight='balanced'` с **upsampling**'ом

In [23]:
def upsample(features, target):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    # рассчитаем соотношение классов
    repeat = round(len(target_zeros) / len(target_ones))
        
    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
    features_upsampled, target_upsampled = shuffle(features_upsampled, target_upsampled, random_state=999)
    
    return features_upsampled, target_upsampled

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Отлично! Красивая функция. Ну и здорово что ты понимаешь, что применение функций упрощает код, делает его структурированным, простым для понимания и упрощает его дальнейшее переиспользование (принцип DRY (Don’t Repeat Yourself / Не повторяйся) – в действии). Дополнительно о DRY можно почитать здесь: https://habr.com/ru/company/itelma/blog/546372</div>

In [24]:
features_train_upsampld, features_target_upsampld = upsample(features_train, target_train)

In [25]:
# проверяем баланс классов
features_target_upsampld.value_counts()

1    5788
0    5553
Name: Exited, dtype: int64

Теперь мы имеем примерно поровну значений в каждом классе. Обучим модель:

In [26]:
# RandomForestClassifier
# сначала подбираем лучшую глубину
best_f1 = 0
best_depth = 0
for depth in range(1, 16):
    model = RandomForestClassifier(random_state=999, max_depth=depth)
    model.fit(features_train_upsampld, features_target_upsampld)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > best_f1:
        best_f1 = f1
        best_depth = depth

# по лучшей глубине подберем лучшее число деревьев
best_estims = 100 #by default
for estims in range(10, 161, 10):
    model = RandomForestClassifier(random_state=999, max_depth=best_depth, n_estimators=estims)
    model.fit(features_train_upsampld, features_target_upsampld)
    predictions = model.predict(features_valid)
    f1 = f1_score(target_valid, predictions)
    if f1 > best_f1:
        best_f1 = f1
        best_estims = estims
        
print('Best depth:', best_depth, 'Best estims', best_estims, 'F1:', best_f1)

Best depth: 11 Best estims 70 F1: 0.6397608370702541


Upsampling показал худшие результаты (0.64) против class_weight='balanced' (0.652)

**Вывод:**
- Лучший F1 показала модель `RandomForestClassifier`
- Были подобраны следующие лучшие гиперпараметры: `depth: 9`, `n_estimators: 120`, `class_weight='balanced'`
- Метрика **F1: 0.652** (на валидационной выборке)
- С несбалансированностью классов лучше помог справиться гиперпараметр `class_weight='balanced'`, нежели `upsampling` обучающих выборок.

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Здесь отлично: подобрали для всех наших моделей лучшие гиперпараметры (в данном случае - максимизирующие f1-меру и(?) AUC ("площадь"). Также здесь мы ещё и определили САМУЮ лучшую модель. На валидации ею оказалась модель "случайного леса". 

После того, как гиперпараметры на валидации подобраны - тестируем модели на тестовых данных. По результатам тестирования на тесте (сорри за тавталогию) выбираем модель, которую сможем передать в продакшн. 
</div>

## Тестирование модели

Заново обучим нашу лучшую модель, т.к. ранее она "перезаписалась"

In [28]:
model = RandomForestClassifier(random_state=999, max_depth=9, n_estimators=120, class_weight='balanced')
model.fit(features_train, target_train)

RandomForestClassifier(class_weight='balanced', max_depth=9, n_estimators=120,
                       random_state=999)

Проверим предсказания на тестовой выборке

In [29]:
predictions = model.predict(features_test)
f1 = f1_score(target_test, predictions)

In [30]:
f1

0.6160990712074303

Посчитаем **ROC-AUC**

In [31]:
prob_test = model.predict_proba(features_test)
prob_test = prob_test[:, 1]

auc_roc = roc_auc_score(target_test, prob_test)
print(auc_roc)

0.8600880802830121


<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Здесь хорошо. Расчёт ROC-AUC выполнен верно, через предсказанную вероятности класса 1. 
    
Площадь под ROC-кривой (Area Under Curve – площадь под кривой, Receiver Operating Characteristic – рабочая характеристика приёмника ... пришло из области радиоэлектроники) – это метрика оценки для задач бинарной классификации. ROC-кривая показывает зависимость между долей верно классифицированных объектов положительного класса и долей ложноположительных объектов. Ну и сама площадь под ROC-кривой является мерой качества классификатора: чем выше значение AUC (Area Under Curve), тем лучше качество классификации конкретной модели. Наш глаз и мозг устроены так, что мы очень хорошо видим геометрическую разницу (в данном случае площадей).
</div>

**Финальный вывод:**
- Данные были открыты, предобработаны, проверены
- Подготовлены обучающая, валидационная и тестовая выборки с примененим техник one-hot encoding, переменные отмасштабированы
- Исследован баланс классов. Выявлен дисбаланс. Модели были обучены с учетом дисбаланса и без. Проведено сравнение работы гиперпараметра `class_weight='balanced'` и техники `upsampling`. Гиперпараметр оказался эффективнее.
- Выяснено, что лучшая модель по метрике F1 score - Это `RandomForestClassifier`. Для неё перебором подобраны наилучшие гиперпараметры.
- **F1** модели на **тестовой** выборке составил `0.618`. **ROC-AUC** score = `0.86`, что говорит о том, что модель предсказывает лучше, чем случайные числа.

<div class="alert alert-success">
<b>ОТЛИЧНО! 👍</b>

Всё отлично, результат достигнут.
    
Один совет (также, на будущее): как попытаться улучшить полученный результат, с минимум усилий? Ответ: мы использовали train для обучения модели, а valid - для поиска лучших значений гиперпараметров. Лучшие параметры нашли. Так почему бы теперь наши модели с выбранными гиперпараметрами не обучить на *общей* (тренировочной + валидационной) выборке (pd.concat() можно использовать для объединения). Чем больше данных, тем *лучше* модели смогут обучиться (надо проверять!). И вот теперь эту дообученную модель мы уже финально проверим на тестовой выборке (test).

Но следует учесть вот какой момент: нужно быть аккуратным с подобным «улучшением», если мы кодируем или масштабируем наши выборки. Например, в следующем проекте мы обучаемся на train’е, а затем делаем transform на валидации и тесте. Если после этого объединить трейн и валид, то это будет не совсем верно.
</div>

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: задача исследована
    - [x]  Исследован баланс классов
    - [x]  Изучены модели без учёта дисбаланса
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 3: учтён дисбаланс
    - [x]  Применено несколько способов борьбы с дисбалансом
    - [x]  Написаны выводы по результатам исследования
- [x]  Выполнен шаг 4: проведено тестирование
- [x]  Удалось достичь *F1*-меры не менее 0.59
- [x]  Исследована метрика *AUC-ROC*

# Результат ревью:

Я отмечаю адекватный уровень твоей программистской подготовки. Ты используешь продвинутые концепции языка Python, библиотеки pandas и одной из основных библиотек машинного обучения sckit-learn. Всё это в целом упрощает твой код и одновременно показывает твой уровень владения инструментами анализа данных и machine learning.

С комментированием никаких проблем нет. Твоих объяснений достаточно для понимания твоими коллегами хода мыслей. Ссылки из брифа проекта в нужных местах также помогают следовать за развитием событий и дают полную информацию касаемо того, что будет происходить в том или ином разделе проекта.

Мне было просто проверять твой проект, потому что он чётко структурирован, с использованием оглавления. 

Зелёным цветом я отметил и прокомментировал удачные и элегантные решения, на которые ты можешь опираться в будущих проектах.

Жёлтым цветом я выделил то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас.

Вадим, у тебя хорошая работа, все четко, осмысленно. Выводы присутствуют, с комментированием кода тоже никаких проблем нет.

Я оставил ряд советов на будущее (на результат этого проекта они не влияют). 

Подводя итог - отличная работа, молодец! Я отправляю тебе проект ещё раз лишь для того, чтобы удостовериться что у тебя нет вопросов ко мне. Если их нет - пошли мне тетрадку ещё раз, и я тут же, без слов приму твою работу!